core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c

```diff
@@ -969,9 +1006,49 @@ wasi_poll_oneoff(wasm_exec_env_t exec_env, const wasi_subscription_t *in,
         || !validate_native_addr(nevents_app, sizeof(uint32)))
         return (wasi_errno_t)-1;
 
-    err = wasmtime_ssp_poll_oneoff(curfds, in, out, nsubscriptions, &nevents);
-    if (err)
-        return err;
+    __wasi_timestamp_t elapsed = 0;
+    const __wasi_timestamp_t timeout = get_timeout_for_poll_oneoff(
+                                 in, nsubscriptions),
+                             time_quant = 1e9;
+    const uint64 size_to_copy = nsubscriptions * sizeof(wasi_subscription_t);
+    __wasi_subscription_t *in_copy = NULL;
+    if (size_to_copy >= UINT32_MAX
+        || (size_to_copy
+            && !(in_copy = (__wasi_subscription_t *)wasm_runtime_malloc(
+                     (uint32)size_to_copy)))) {
+        return (wasi_errno_t)-1;
+    }
+    bh_memcpy_s(in_copy, size_to_copy, in, size_to_copy);
+
+    while (1) {
```

Comment:
This is just a personal opinion so feel free to ignore - I'm not a big fan of infinite loops (`while(1)`) and prefer to avoid them; instead, I prefer have a clear stop condition in the loop declaration. In this case, we could have:
```
while (timeout != (__wasi_timestamp_t)-1 && elapsed > timeout)
```

In fact, that change actually fixes the bug - currently, if the initial `timeout` is smaller than `time_quant`, we'll never execute `wasmtime_ssp_poll_oneoff`. I think we should always wait for the reminder of `timeout % time_quant`.

Reply:
but don't we want to loop infinitely if timeout is -1?

Reply:
i think it should instead be the ! of the above, changing it